Folder structure validation

qiaozha opened the pull request:
## Folder Structure Validation

This PR introduces folder structure validation to enforce Azure REST API specifications organization standards.

### Summary

Adds a new independent `folder-structure` validation check that runs alongside TypeSpec validation to ensure specifications follow the v2 folder structure guidelines. This helps maintain consistency across the repository and prevents introducing legacy v1 folder structures in v2-compliant organizations.

### Key Changes

#### 1. New Folder Structure Validation Tool
- **New PowerShell script**: `eng/scripts/Folder-Structure-Validation.ps1` - Validates `specification/{org}` folders for any changed files
- **New CLI tool**: `npx folder-structure` - Validates folder structure at the org level
- **Comprehensive validation logic**: 
  - Detects v1 vs v2 folder structures
  - Prevents v1 structure regression in v2-compliant orgs
  - Allows existing v1 services in mixed orgs but blocks new v1 services
  - Provides detailed reporting of v1-like and v2-aligned folders

#### 2. GitHub Actions Workflow Updates
- Renamed `typespec-validation.yaml` → `spec-validation.yaml` with two independent jobs:
  - **TypeSpec Validation**: Runs `npx tsv` on TypeSpec folders
  - **Folder Structure Validation**: Runs `npx folder-structure` on org folders
- **TypeSpec Validation - All** workflow unchanged (only validates TypeSpec compilation)

#### 3. Tool Reorganization
- Renamed `eng/tools/typespec-validation` → `eng/tools/spec-validation`
- Package name: `@azure-tools/typespec-validation` → `@azure-tools/spec-validation`
- Two CLI commands:
  - `npx tsv` - TypeSpec validation (compilation, linting, SDK config)
  - `npx folder-structure` - Folder structure validation

#### 4. Suppression Updates
- **Tool name standardization**: Changed all suppressions from `tool: TypeSpecValidation` to `tool: SpecValidation`
  - Both `tsv` and `folder-structure` tools now use the same tool name for suppressions
- **New suppression**: Added `FolderStructure` rule suppression for `specification/compute` (legacy structure)

#### 5. Enhanced TypeSpec Folder Detection
- Updated `Get-TypeSpec-Folders.ps1` to return:
  - `$orgFolders` - for folder-structure validation (all `specification/{org}` folders)
  - `$typespecFolders` - for TypeSpec validation (folders with `tspconfig.yaml`)
  - `$checkingAllSpecs` - flag for validation mode

### Validation Behavior

#### Regular Spec Validation (PR checks)
- Runs `npx folder-structure` on all `specification/{org}` folders with changes
- Runs `npx tsv` on TypeSpec projects (folders with `tspconfig.yaml`)

#### TypeSpec Validation - All (scheduled/manual)
- Only runs `npx tsv` on TypeSpec projects
- Skips folder-structure validation

### Folder Structure Rules

**V2 Structure** (enforced for new services):
```
specification/{org}/data-plane/{ServiceName}/...
specification/{org}/resource-manager/{RPNamespace}/{ServiceName}/...
```

**V1 Structure** (legacy, blocked for new services):
```
specification/{org}/{ServiceFolder}/...
specification/{org}/resource-manager/{RPNamespace}/... // (without {ServiceName} layer)
```

**Validation Logic**:
- Pure v2 orgs: Must not introduce any v1 structure
- Mixed orgs: Can modify existing v1 services, but cannot add new v1 services
- V1 orgs: No restrictions (for now)

### Testing
- Comprehensive test suite in `eng/tools/spec-validation/test/folder-structure.test.ts`
- Tests cover v1/v2 detection, regression prevention, and various edge cases

### Documentation
- Updated `documentation/ci-fix.md` with folder structure validation information
- Updated GitHub workflow summaries and check names

---

**Breaking Changes**: None - existing specs continue to work with suppressions where needed

**Migration Path**: Services using v1 structure can continue operating but should plan migration to v2 structure

qiaozha committed:
folder-structure-validation

qiaozha committed:
fix folder structure validation logic

qiaozha commented:
> @qiaozha: My understanding is "folder structure v2" is only a concept we care about enforcing for TypeSpec specs. If a swagger-only spec wants to use the same folder and readme structure, they can, but this technically isn't "folder structure v2". And we're not interested in adding more checks or tooling for swagger-only specs.
> 
> Given this, can the check be added to existing TypeSpecValidation, instead of needing a whole new tool and workflow?

Folder structure v2 is not just about TypeSpec, it also has impacts on the services with pure swaggers as well. Just like the service version uniform guideline are not bound by whether that service is using typespec or swagger, they need to deliver service that are versioning uniformly regardless. And folder structure v2 is the fundamental support of the version uniform guideline. @JeffreyRichter feel free to correct me if I am wrong. 

You can take a look at this https://github.com/Azure/azure-rest-api-specs/tree/main/specification/orbital/resource-manager/Microsoft.Orbital example, it doesn't have mgmt plane TypeSpec, but this service is trying to isolate as two different services in their mgmt plane with this weird way.  After the service team sort this out by adopting to our folder structure v2 like to have two serviceName folders under Microsoft.Orbital namespace folder, when they are future PRs trying to add a new stable or preview version under Microsoft.Orbital directly, we are definitely not going to allow this.

mikeharder commented:
> Folder structure v2 is not just about TypeSpec, it also has impacts on the services with pure swaggers as well.

Azure SDK EngSys has been instructed to focus new feature work only on TypeSpec specs.  We need to continue supporting swagger-only specs (e.g. fixing bugs in swagger-based tools), but all new tools and checks (like enforcing folder structure v2) should only apply to TypeSpec specs.

> You can take a look at this https://github.com/Azure/azure-rest-api-specs/tree/main/specification/orbital/resource-manager/Microsoft.Orbital example, it doesn't have mgmt plane TypeSpec, but this service is trying to isolate as two different services in their mgmt plane with this weird way. After the service team sort this out by adopting to our folder structure v2 like to have two serviceName folders under Microsoft.Orbital namespace folder, when they are future PRs trying to add a new stable or preview version under Microsoft.Orbital directly, we are definitely not going to allow this.

Instead of adding a new tool and check for a small number of swagger-only specs converting to something like v2 (but without TypeSpec), I think the effort would better be spent, accelerating the conversion of these specs to TypeSpec.

> Another consideration is, I am not sure if you remember that I pointed it out to you before, the currently way where we put the folder structure validation inside the TypeSpec validation is not clear enough for service team to focus on which failure to fix first when they see a lot of CI failures. But I think folder structure is one of the fundamental checks that they should fix first if there's any violations.

Rather than adding a new tool and check, the better fix would be to make errors from TypeSpecValidation easier to read (e.g. add a markdown summary), either for all types of errors, or just folder structure.

qiaozha commented:
@mikeharder 
I don't think swagger only services is a small number. From my previous statistics, it has more than 110 services. The overall communicate will be very huge if we let them do whatever they want after we helped them migrate to folder structure v2. 
Compared to the communication effort, I don't think add a new check will take that much effort and even less from EngSys team as I am taking the coding part, right? 

If we add this check within TypeSpec validation, this will be confusing to those 110+ service teams if they don't have any typespec at all. And we will probably get a lot of questions from them, right?

Even if we don't invest in swagger, but I don't think some minimal fundamental tooling support to make the SDK engineer life easier is not something we should do. 

Also, do you think such kind of questions might be better to discuss in the email thread instead of PR?

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
rename to spec validation

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
merge folder structure validation into spec validation

qiaozha committed:
merge main

qiaozha committed:
fix tests and eslint

qiaozha committed:
merge main

qiaozha committed:
fix eslint

qiaozha committed:
fix test

qiaozha committed:
fix format

qiaozha committed:
merge main

qiaozha committed:
Fix mock for v2 compliance enforcement test

The test was failing in CI because the git ls-tree mock was not properly matching the actual command structure. The real command uses a format like 'origin/branch:path' but the mock was only checking for the branch name substring.

This fix makes the mock more robust by properly parsing the git ls-tree command arguments to match the expected branch and path structure.

qiaozha committed:
Make folder structure test more robust for CI environments

- Improve git ls-tree mock to handle different argument formats
- Add explicit detached flag for CI branch scenario compatibility
- Handle potential whitespace variations in path parsing
- Add defensive checks for multiple argument patterns

This addresses potential CI environment differences where git commands
might have slightly different formatting or execution context.

qiaozha committed:
fix format

qiaozha committed:
Improve v2 compliance test robustness and error handling

Key improvements:
- Simplify git mock with broader pattern matching for CI compatibility
- Use more specific error message assertion for better test reliability
- Add test case for git operation failures to ensure graceful fallback
- Remove debug logging for cleaner output

This addresses potential CI environment differences by making the test
more flexible while maintaining the same validation logic.

qiaozha committed:
fix format

qiaozha committed:
fix tests

qiaozha committed:
Fix lint errors: convert error to string in template literal and remove unnecessary async keywords

- Fix TypeScript error in folder-structure.ts by converting error to String(error) in template literal
- Remove unnecessary async keywords from beforeEach functions that don't use await
- All lint checks now pass clean
- All tests continue to pass (33/33)

qiaozha committed:
refactor: unify test mocking approach in folder-structure tests

- Remove all direct process.env.GITHUB_BASE_REF manipulation from tests
- Consolidate on git mocking for consistent target branch detection simulation
- Eliminate confusing mixed approach of environment variables + git mocks
- Maintain all 33 tests passing with cleaner, more maintainable test code

qiaozha committed:
fix: improve CI scenario test mock for upstream remote target branch detection

- Fix git mock to properly handle target branch detection in CI environments
- Add more robust handling of rev-parse and ls-tree commands for upstream remotes
- Ensure test properly simulates main branch as target with v2 structure indicators
- Test now correctly fails validation when target branch has v2 structure

qiaozha committed:
fix format

qiaozha committed:
revert: fix upstream CI scenario test to use original working mock

- Restore original target branch detection logic
- Remove overly broad git mock fallbacks that interfered with other tests
- Keep test focused on upstream/feature/v2-structure scenario
- Ensures all 33 tests pass without interference

qiaozha committed:
fix: prevent git mock state pollution between test suites

- Add mock isolation to v2 compliance enforcement tests
- Save/restore original mock implementation in beforeEach/afterEach
- Prevent test failures caused by mock state bleeding between describe blocks
- Ensures tests run independently regardless of execution order

Fixes the issue where commit d291ba64d4 caused other tests to fail
due to mock state pollution from the upstream CI scenario test.

qiaozha committed:
Fix TypeScript linting errors in folder-structure tests

- Replace explicit 'any' types with 'unknown' types in mock restoration
- Resolve @typescript-eslint/no-explicit-any violations
- Maintain type safety while satisfying linting rules

qiaozha committed:
update shouldEnforceV2Compliance logic

qiaozha committed:
fix env issue

qiaozha committed:
try fix ci failure

qiaozha committed:
update folder structure as independent rule

qiaozha committed:
support npx folder-structure

qiaozha committed:
feat(folder-structure): improve --validate reporting and fix -f flag handling

- Remove -f flag requirement; folder can be passed as positional argument
- Fix bug where -f flag was not being parsed correctly (was running PR mode instead)
- Change --validate mode to analyze and report folder structure compliance
- Group folders at preview/stable level to avoid drowning in API version subfolders
- Show which folders are v2-aligned vs v1-like with clear visual markers
- Report overall org structure (v1/v2/mixed) and exit appropriately

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
fix(folder-structure): collapse examples folders and fix v1 detection

- Group files under /examples/<apiVersion>/ into single examples folder unit
- Fix bug where examples folders were incorrectly marked as v1-like
- All truncated folder paths now go through same v1/v2 detection logic
- Cleaner output with preview/stable/examples collapsed appropriately

qiaozha committed:
fix(folder-structure): require proper path depth for v2 classification

- Data-plane paths need at least 4 parts (specification/org/data-plane/ServiceName)
- Resource-manager paths need at least 5 parts (specification/org/resource-manager/RP/ServiceName)
- Shallow parent paths like 'specification/storage/resource-manager' are now correctly classified as 'other' instead of v2-aligned
- Fixes incorrect v2 classification for intermediate directory paths

qiaozha committed:
fix(folder-structure): ignore file extensions when checking for namespace prefixes

- Files like suppressions.yaml were incorrectly flagged as v1 due to dots in filename
- Now only check for dots in directory names (service names), not file names
- Fixes false positive v1 detection for orgs like maps

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
Revert typespec-validation-all to TypeSpec-only validation

- Rename spec-validation-all.yaml back to typespec-validation-all.yaml
- Remove folder-structure validation from TypeSpec-Validation.ps1
- Restore TypeSpec-specific naming and messages
- Keep TypeSpec Validation - All check focused on TypeSpec only

qiaozha committed:
Revert documentation links from spec-validation to typespec-validation

- Restore https://aka.ms/azsdk/specs/typespec-validation links
- Update sdk-tspconfig-validation.ts
- Update tspconfig.yaml files in edge and edgemarketplace specs

qiaozha committed:
update workflow name and docs etc

msyyc committed:
fix python config

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
update suppression to spec validation

qiaozha committed:
add folder-structure as a step check in spec-validation

qiaozha committed:
add folder-structure suppresion for compute

ravimeda committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
verify the impacted specs

qiaozha committed:
Merge branch 'main' into folder-structure-validation

qiaozha committed:
Merge branch 'main' into folder-structure-validation
